### PR TITLE
Add validation frequency option 

### PR DESCRIPTION
**Pull request by melihyilmaz**
I put together this PR after coming to the conclusion that there was no option to run validation before a training epoch ends, which is crucial when training on large datasets like MSKB, but it also includes fixes to issues with logging and checkpointing we introduced earlier (1st commit: e11b80ada0e66f62a64fa880a33bdeb3865d99d1). To list the changes:

- `every_n_train_steps` option now controls both the frequency of validation epochs and model checkpointing. We can decouple the two but kept it as such for simplicity.
-  Validation performance metrics are logged (and added to tensorboard) at the validation epoch, and training loss is logged at the end of training epoch, i.e. training and validation metrics are logged asynchronously.
- We now log `step` rather than `epoch` as the unit of progress.
- Fixed `tb_summarywriter` not being read of the `config.yaml`.
- Fixed metrics being logged, discontinued legacy metric `recall`.
- Fixed `save_model` option, previously didn't have any effect. 

I tested the changes with different settings and things seem to work but our unit tests still don't cover the changed functions.

### CODE REVIEW

File: casanovo/config.yaml [hunk 1]
```diff
@@ -117,7 +117,7 @@ save_model: True
 model_save_folder_path: ""
 # Set to "False" to save the PyTorch model instance
 save_weights_only: True
-# Model checkpointing frequency in training steps
+# Model validation and checkpointing frequency in training steps
```
Comment (wfondrie) [inline casanovo/config.yaml:120]: **question (non-blocking):** Do we want separate parameters for these? I haven't been training models so I don't know if this is sufficient or if they should be split.

**Review by wfondrie — APPROVED** [review]
This looks good to me. I left one non-blocking question though. You should also make sure to note this in the changelog.

File: casanovo/config.yaml (hunk 1 above)
Comment (melihyilmaz) [reply casanovo/config.yaml:120]: Separation would give flexibility to cases users who don't want to checkpoint as often as the validation but I think this keeps things simpler and works fine.

File: casanovo/denovo/model.py [hunk 2]
```diff
@@ -890,36 +894,35 @@ def _log_history(self) -> None:
...
+                        ("eval/val_pep_precision", "valid_pep_precision"),
+                        ("eval/val_aa_precision", "valid_aa_precision"),
                     ]:
-                        self.tb_summarywriter.add_scalar(
-                            descr,
-                            metrics.get(key, np.nan),
-                            metrics["epoch"] + 1,
-                        )
+                        metric_value = metrics.get(key, np.nan)
+                        if metric_value is not np.nan:
```
Comment (bittremieux) [inline casanovo/denovo/model.py:920]: **issue:** This is not fully correct, you should use `np.isnan` instead.

File: casanovo/denovo/model.py [hunk 3]
```diff
@@ -890,36 +894,35 @@ def _log_history(self) -> None:
...
-                    "Epoch\tTrain loss\tValid loss\tAA precision\tAA recall\t"
-                    "Peptide recall"
-                )
-            metrics = self._history[-1]
-            if metrics["epoch"] % self.n_log == 0:
+        if len(self._history) == 1:
+            logger.info(
+                "Step\tTrain loss\tValid loss\tPeptide precision\tAA precision"
+            )
+        metrics = self._history[-1]
```
Comment (bittremieux) [inline casanovo/denovo/model.py:901]: **issue:** This needs to go after making sure that the log is not empty.
